kaldb/src/main/java/com/slack/kaldb/clusterManager/SnapshotDeletionService.java

```diff
@@ -0,0 +1,212 @@
+
+  public SnapshotDeletionService(
+      ReplicaMetadataStore replicaMetadataStore,
+      SnapshotMetadataStore snapshotMetadataStore,
+      S3BlobFs s3BlobFs,
+      KaldbConfigs.ManagerConfig managerConfig,
+      MeterRegistry meterRegistry) {
+
+    checkArgument(
+        managerConfig.getSnapshotDeletionServiceConfig().getSnapshotLifespanMins()
+            > managerConfig.getReplicaCreationServiceConfig().getReplicaLifespanMins(),
+        "SnapshotLifespanMins must be greater than the ReplicaLifespanMins");
+    // schedule configs checked as part of the AbstractScheduledService
+
+    this.managerConfig = managerConfig;
+    this.replicaMetadataStore = replicaMetadataStore;
+    this.snapshotMetadataStore = snapshotMetadataStore;
+    this.s3BlobFs = s3BlobFs;
+    this.meterRegistry = meterRegistry;
+
+    this.snapshotDeleteSuccess = meterRegistry.counter(SNAPSHOT_DELETE_SUCCESS);
+    this.snapshotDeleteFailed = meterRegistry.counter(SNAPSHOT_DELETE_FAILED);
+    this.snapshotDeleteTimer = meterRegistry.timer(SNAPSHOT_DELETE_TIMER);
+  }
+
+  @Override
+  protected void runOneIteration() {
+    deleteExpiredSnapshotsWithoutReplicas();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(
+        managerConfig.getScheduleInitialDelayMins(),
+        managerConfig.getSnapshotDeletionServiceConfig().getSchedulePeriodMins(),
+        TimeUnit.MINUTES);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting snapshot deletion service");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Closing snapshot deletion service");
+    executorService.shutdown();
+  }
+
+  /**
+   * Deletes snapshots and associated object storage data that have no corresponding replicas.
+   * Separate services are responsible for the eviction and deletion of expired replicas, and this
+   * service only is expected to remove snapshots that are not being currently served.
+   *
+   * <p>Consideration should be taken for when configs are changed, such as shortening the replica
+   * lifespan. The expectation is that previously created replicas that may have a longer expiration
+   * would see that respected throughout the duration of their lifespan.
+   *
+   * <p>Deletion of expired snapshots should gracefully handle the scenario where immediately prior
+   * to reaching the end of the configured snapshot lifespan, an on-demand request is received. In
+   * this scenario a new replica would have been created, and the object would be currently
+   * downloading for servicing the request. The expectation from the end user would be that the
+   * request would still be served, as when the request was made it was still "in compliance" with
+   * the lifespan configuration.
+   *
+   * <p>When creating on-demand replicas, care should be taken to not create replicas for snapshots
```

Comment:
To minimize the race conditions, I think it's best to delete an object a few more hours after expiration. This value can be hardcoded for now to 6 hours. In future, we can leave a TODO to make this value configurable.

Reply:
This isn't needed, as snapshots already have a configurable expiration (`snapshotLifespanMins: 10080`) that is significantly longer than that of replicas (`replicaLifespanMins: 1440`). This note is simply to call out that new replicas should not be created for expired snapshots that are pending deletion. 

This is something we should check for when making an on-demand request for a snapshot (ie, snapshot is not expired) - since this code isn't written yet though I went ahead and added a note here.

Reply:
Done.